### PR TITLE
Fix parsing of JSON objects in BrowserCacheManager.ts

### LINKED ISSUE #2521: SyntaxError: Unexpected token e in JSON at position 0 at JSON.parse (<anonymous>) at t.getIdTokenCredential
>> Please follow the issue template below. Failure to do so will result in a delay in answering your question.

## Library
- [ ] `msal@1.x.x` or `@azure/msal@1.x.x`
- [X] `@azure/msal-browser@2.4.1`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

**Important**: Please fill in your exact version number above, e.g. `msal@1.1.3`.

## Framework
React
## Description
Got a syntax error while trying to login
Uncaught (in promise) SyntaxError: Unexpected token e in JSON at position 0
    at JSON.parse (<anonymous>)
    at t.getIdTokenCredential (index.d9b6c51d.js:2)
    at t.e.getSpecificCredential (index.d9b6c51d.js:2)
    at index.d9b6c51d.js:2
    at Array.forEach (<anonymous>)
    at t.e.getCredentialsFilteredByInternal (index.d9b6c51d.js:2)
    at t.e.getCredentialsFilteredBy (index.d9b6c51d.js:2)
    at t.e.readRefreshTokenFromCache (index.d9b6c51d.js:2)
    at t.<anonymous> (index.d9b6c51d.js:2)
    at index.d9b6c51d.js:2

## Error Message
Uncaught (in promise) SyntaxError: Unexpected token e in JSON at position 0
    at JSON.parse (<anonymous>)
    at t.getIdTokenCredential (index.d9b6c51d.js:2)
    at t.e.getSpecificCredential (index.d9b6c51d.js:2)
    at index.d9b6c51d.js:2
    at Array.forEach (<anonymous>)
    at t.e.getCredentialsFilteredByInternal (index.d9b6c51d.js:2)
    at t.e.getCredentialsFilteredBy (index.d9b6c51d.js:2)
    at t.e.readRefreshTokenFromCache (index.d9b6c51d.js:2)
    at t.<anonymous> (index.d9b6c51d.js:2)
    at index.d9b6c51d.js:2
## Security
- [ ] Is this issue security related?

## Regression
- [X] Did this behavior work before? 
Version: 2.2.0
I think so.. this is the first time I've seen this error since upgrading to 2.4.1 from 2.2.0
## MSAL Configuration
```js
    const publicClientApp = new PublicClientApplication({
      auth: {
        clientId: this.clientId,
        authority: this.authority,
        redirectUri: this.loginRedirectUri,
        postLogoutRedirectUri: this.postLogoutRedirectUri
      },
      cache: {
        cacheLocation: "localStorage",
        storeAuthStateInCookie: isIEOrOldEdge // https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/1188#issuecomment-634281307
      },
      system: {
        tokenRenewalOffsetSeconds: 60,
        iframeHashTimeout: 30000, // iframe is used to refresh the refresh_token as it has a 24h lifetime
        loggerOptions: {
          loggerCallback: MsalAuthContext.msalLoggerCallback,
          piiLoggingEnabled: false,
          logLevel: LogLevel.Verbose
        }
      }
    });
```

## Reproduction steps
```js
// Provide relevant code snippets here.
// For Azure B2C issues, please include your policies.
```

## Expected behavior
No error
## Browsers/Environment
- [X] Chrome
- [ ] Firefox
- [ ] Edge
- [ ] Safari
- [ ] IE
- [ ] Other (Please add browser name here)

### PR DESCRIPTION
**Pull request by pkanher617**
This PR addresses the issue raised in #2521 and also makes the same fixes done in 1.x to fix JSON cache parsing (#2510).

### CODE REVIEW

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 1]
```diff
@@ -464,7 +487,7 @@ export class BrowserCacheManager extends CacheManager {
     generateCacheKey(key: string): string {
         try {
             // Defined schemas do not need the key migrated
-            this.IsJSON(key);
+            this.validateAndParseJson(key);
             return key;
```
Comment (jasonnutter) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:491]:
```suggestion
            return this.validateAndParseJson(key);
```

**Review by jasonnutter — APPROVED** [review]

File: change/@azure-msal-browser-2020-10-30-14-51-02-cache-fix-json-parse-v2.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "patch",
+  "comment": "Fix JSON.parse issue and cache value validation (#2527)",
+  "packageName": "@azure/msal-browser",
+  "email": "prkanher@microsoft.com",
+  "dependentChangeType": "patch",
```
Comment (sameerag) [inline change/@azure-msal-browser-2020-10-30-14-51-02-cache-fix-json-parse-v2.json:6]:
```suggestion
  "dependentChangeType": "none",
```

**Review by sameerag — APPROVED** [review]

**Comment by sameerag** [comment]
Can we add unit tests?

File: change/@azure-msal-browser-2020-10-30-14-51-02-cache-fix-json-parse-v2.json (hunk 2 above)
Comment (pkanher617) [reply change/@azure-msal-browser-2020-10-30-14-51-02-cache-fix-json-parse-v2.json:6]: shouldn't be an issue with #2528

File: lib/msal-browser/src/cache/BrowserCacheManager.ts [hunk 3]
```diff
@@ -103,8 +103,19 @@ export class BrowserCacheManager extends CacheManager {
      * Parses passed value as JSON object, JSON.parse() will throw an error.
      * @param input
      */
-    private IsJSON(jsonValue: string): void {
-        JSON.parse(jsonValue);
+    private validateAndParseJson(jsonValue: string): object {
```
Comment (jmckennon) [inline lib/msal-browser/src/cache/BrowserCacheManager.ts:106]: should this be in `common` somewhere? do other libraries also want access to something like this?

**Review by jmckennon — APPROVED** [review]

File: lib/msal-browser/src/cache/BrowserCacheManager.ts (hunk 3 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts:106]: I think it's somewhat specific to the way the browser handles cache, so I don't think we need to bring it up to the common lib. If we think it is necessary in the future happy to revisit at that time.

Comment (jmckennon) [reply lib/msal-browser/src/cache/BrowserCacheManager.ts:106]: sounds good then!

**Pull request merged**